### PR TITLE
Add tests for self.queueMicrotask()

### PR DESCRIPTION
**Pull request by domenic**
Follows https://github.com/whatwg/html/pull/2789.

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246403042)

# Firefox (nightly) #
Testing web-platform-tests at revision 70b25df06540ae2d26c182be4f5d6897bf43fa08
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.html">/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.html</a></summary>


|         Subtest          | Results |             Messages            |
|--------------------------|---------|---------------------------------|
|                          | OK      |                                 |
| `It rethrows exceptions` | FAIL    | `queueMicrotask is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.worker.html">/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.worker.html</a></summary>


|         Subtest          | Results |             Messages            |
|--------------------------|---------|---------------------------------|
|                          | OK      |                                 |
| `It rethrows exceptions` | FAIL    | `queueMicrotask is not defined` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.any.html">/html/webappapis/microtask-queuing/queue-microtask.any.html</a></summary>


|                  Subtest                   | Results |                                                                                         Messages                                                                                         |
|--------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                            | OK      |                                                                                                                                                                                          |
| `It exists and is a function`              | FAIL    | `assert_equals: expected "function" but got "undefined"`                                                                                                                                 |
| `It throws when given non-functions`       | FAIL    | `assert_throws: no argument function "() => queueMicrotask()" threw object "ReferenceError: queueMicrotask is not defined" ("ReferenceError") expected object "TypeError" ("TypeError")` |
| `It calls the callback asynchronously`     | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |
| `It does not pass any arguments`           | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |
| `It interleaves with promises as expected` | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.any.worker.html">/html/webappapis/microtask-queuing/queue-microtask.any.worker.html</a></summary>


|                  Subtest                   | Results |                                                                                         Messages                                                                                         |
|--------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                            | OK      |                                                                                                                                                                                          |
| `It exists and is a function`              | FAIL    | `assert_equals: expected "function" but got "undefined"`                                                                                                                                 |
| `It throws when given non-functions`       | FAIL    | `assert_throws: no argument function "() => queueMicrotask()" threw object "ReferenceError: queueMicrotask is not defined" ("ReferenceError") expected object "TypeError" ("TypeError")` |
| `It calls the callback asynchronously`     | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |
| `It does not pass any arguments`           | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |
| `It interleaves with promises as expected` | FAIL    | `queueMicrotask is not defined`                                                                                                                                                          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.window.html">/html/webappapis/microtask-queuing/queue-microtask.window.html</a></summary>


|                                  Subtest                                  | Results |                                                Messages                                                |
|---------------------------------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------|
|                                                                           | OK      |                                                                                                        |
| `It interleaves with MutationObservers as expected`                       | FAIL    | `promise_test: Unhandled rejection with value: object "ReferenceError: queueMicrotask is not defined"` |
| `It interleaves with MutationObservers and promises together as expected` | FAIL    | `promise_test: Unhandled rejection with value: object "ReferenceError: queueMicrotask is not defined"` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246403043)

# Sauce (safari) #
Testing web-platform-tests at revision 70b25df06540ae2d26c182be4f5d6897bf43fa08
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>5 tests ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.html">/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.html</a></summary>


|         Subtest          | Results |                Messages               |
|--------------------------|---------|---------------------------------------|
|                          | OK      |                                       |
| `It rethrows exceptions` | FAIL    | `Can't find variable: queueMicrotask` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.worker.html">/html/webappapis/microtask-queuing/queue-microtask-exceptions.any.worker.html</a></summary>


|         Subtest          | Results |                Messages               |
|--------------------------|---------|---------------------------------------|
|                          | OK      |                                       |
| `It rethrows exceptions` | FAIL    | `Can't find variable: queueMicrotask` |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.any.html">/html/webappapis/microtask-queuing/queue-microtask.any.html</a></summary>


|                  Subtest                   | Results |                                                                                            Messages                                                                                            |
|--------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                            | OK      |                                                                                                                                                                                                |
| `It exists and is a function`              | FAIL    | `assert_equals: expected "function" but got "undefined"`                                                                                                                                       |
| `It throws when given non-functions`       | FAIL    | `assert_throws: no argument function "() => queueMicrotask()" threw object "ReferenceError: Can't find variable: queueMicrotask" ("ReferenceError") expected object "TypeError" ("TypeError")` |
| `It calls the callback asynchronously`     | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |
| `It does not pass any arguments`           | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |
| `It interleaves with promises as expected` | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.any.worker.html">/html/webappapis/microtask-queuing/queue-microtask.any.worker.html</a></summary>


|                  Subtest                   | Results |                                                                                            Messages                                                                                            |
|--------------------------------------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                            | OK      |                                                                                                                                                                                                |
| `It exists and is a function`              | FAIL    | `assert_equals: expected "function" but got "undefined"`                                                                                                                                       |
| `It throws when given non-functions`       | FAIL    | `assert_throws: no argument function "() => queueMicrotask()" threw object "ReferenceError: Can't find variable: queueMicrotask" ("ReferenceError") expected object "TypeError" ("TypeError")` |
| `It calls the callback asynchronously`     | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |
| `It does not pass any arguments`           | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |
| `It interleaves with promises as expected` | FAIL    | `Can't find variable: queueMicrotask`                                                                                                                                                          |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/6334/html/webappapis/microtask-queuing/queue-microtask.window.html">/html/webappapis/microtask-queuing/queue-microtask.window.html</a></summary>


|                                  Subtest                                  | Results |                                                   Messages                                                   |
|---------------------------------------------------------------------------|---------|--------------------------------------------------------------------------------------------------------------|
|                                                                           | OK      |                                                                                                              |
| `It interleaves with MutationObservers as expected`                       | FAIL    | `promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: queueMicrotask"` |
| `It interleaves with MutationObservers and promises together as expected` | FAIL    | `promise_test: Unhandled rejection with value: object "ReferenceError: Can't find variable: queueMicrotask"` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246403049)

# Chrome (unstable) #
Testing web-platform-tests at revision 70b25df06540ae2d26c182be4f5d6897bf43fa08

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/246403050)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 70b25df06540ae2d26c182be4f5d6897bf43fa08

File: html/webappapis/microtask-queuing/queue-microtask.any.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"use strict";
```
Comment (annevk) [inline html/webappapis/microtask-queuing/queue-microtask.any.js]: Make this also run in shared and service workers perhaps?

File: html/webappapis/microtask-queuing/queue-microtask-exceptions.any.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+"use strict";
+
+setup({
+  allow_uncaught_exception: true
+});
+
+async_test(t => {
+  const error = new Error("boo");
+  self.addEventListener("error", ev => {
```
Comment (annevk) [inline html/webappapis/microtask-queuing/queue-microtask-exceptions.any.js]: Shouldn't this use t.step_func_done?

**Comment by domenic** [comment]
@annevk good catches, review comments addressed.

**Review by annevk — APPROVED** [review]

**Pull request merged**